networkx/algorithms/sampling.py

```diff
@@ -0,0 +1,278 @@
+
+#     Examples
+#     --------
+#     Estimate the fraction of labeled elements in the wheel graph on ten
+#     vertices, where the first five nodes are labeled::
+
+#         >>> import networkx as nx
+#         >>> G = nx.wheel_graph(10)
+#         >>> is_labeled = lambda v: v in range(5)
+#         >>> estimate_relative_size(S, labeled=is_labeled, weights=G.degree)
+#         0.50679999999999814
+
+#     """
+#     values = lambda s: 1 if labeled(s) else 0
+#     return estimate_mean(sample, values, weights)
+
+
+def sample_with_replacements(population, size=-1, weights=None):
+    """Generates random items with replacement from a population.
+
+    This function is an iterator generator.
+
+    Parameters
+    ----------
+    population : list
+       List of elements from which to sample.
+
+    size : int
+       Desired number of samples. If this is ``-1``, the default, then the
+       generator never stops.
+
+    weights : function
+       This must be a function whose inputs are elements from ``population``
+       and whose outputs are :class:`float`s, representing the weight of that
+       element when sampling. A higher weight means the elements is more likely
+       to be sampled.
+
+    Returns
+    -------
+    iterator
+       An iterator over elements of ``population``.
+
+    References
+    ----------
+    .. [1]: Gjoka, Minas, Maciej Kurant, Carter T. Butts,
+            and Athina Markopoulou.
+            "Walking in Facebook: A Case Study of Unbiased Sampling of OSNs."
+            *INFOCOM, 2010 Proceedings IEEE*. IEEE, 2010.
+            http://dx.doi.org/10.1109/INFCOM.2010.5462078
+
+    """
+    if weights is None:
+        for c in itertools.count():
+            if c == size:
+                return
+            yield random.choice(population)
+    cum_weights = nx.utils.accumulate(weights(v) for v in population)
+    tot_weight = cum_weights[-1]
+    for c in itertools.count():
+        if c == size:
+            return
+        i = bisect.bisect_right(cum_weights, random.random() * tot_weight)
+        yield population[i]
+
+
+def uniform_independent_node_sample(G, size=-1):
+    """Generates a uniform independent sample of nodes with replacement.
+
+    Parameters
+    ----------
+    G : a NetworkX graph
+       The graph whose nodes will be sampled.
+
+    size: int
+       The number of nodes to sample from the graph. If size is -1, the
+       default, then the generator never stops.
+
+    Returns
+    -------
+    iterator
+       An iterator over nodes of the graph.
+
+    Examples
+    --------
+    Sampling twenty nodes uniformly at random from the wheel graph on ten
+    vertices::
+
+        >>> import networkx as nx
+        >>> G = nx.generators.wheel_graph(10)
+        >>> list(uniform_independent_node_sample(G, size=20))
+        [8, 6, 4, 7, 9, 5, 5, 2, 4, 5, 7, 9, 3, 6, 1, 6, 9, 8, 6, 4]
+
+    """
+    return sample_with_replacements(G, size)
```

Comment:
The first argument to `sample_with_replacements` should be a `list` as documented.